### ISSUE TITLE
Add `--cache` option for faster linting

### ISSUE THREAD

**Issue by ilearnio** [palantir/tslint#4099]
### Feature request

**Is your feature request that we implement a new rule?**
No

**Is your feature request related to a problem? Please describe.**
We just moved our project from JS + Eslint to TS + Tslint, and now every time we run `tslint --project tsconfig.json` in our project's folder It's taking about 40 sec for not so big project (266 .ts files). Before it was taking 2-4 sec with `eslint --cache`. We run linter on every `git push` so it's a bit annoying to wait every time.

**Describe the solution you'd like**
In `eslint` there is a `--cache` option that when passed will create a `.eslintcache`json file that contains mtime of every file he checked last time. On next run it will not check the file again unless mtime changed.

**Additional context**
The `.eslintcache` file looks like the following:

```js
{
  "/Users/Admin/Code/my-project/foo.js": {
    "size": 2283,
    "mtime": 1533226569740,
    "hashOfConfig": "rlkggo",
    "results": {
      "filePath": "/Users/Admin/Code/my-project/foo.js",
      "messages": [],
      "errorCount": 0,
      "warningCount": 0,
      "fixableErrorCount": 0,
      "fixableWarningCount": 0
    }
  },
  "/Users/Admin/Code/my-project/bar.js": {
    "size": 2449,
    "mtime": 1533058785404,
    "hashOfConfig": "rlkggo",
    "results": {
      "filePath": "/Users/Admin/Code/my-project/bar.js",
      "messages": [],
      "errorCount": 0,
      "warningCount": 0,
      "fixableErrorCount": 0,
      "fixableWarningCount": 0
    },
  },
  ...
```

**Comment by JoshuaKGoldberg** [comment]
This is a little harder with typed rules, as changes to one file may cause lint complaints in another file. For example, adding a `@deprecated` tag can cause other files to complain on the `deprecation` rule.

**Comment by johnwiseheart** [comment]
As @JoshuaKGoldberg  mentions, this is a pretty difficult thing to implement as you would also have to cache a graph of dependencies, such that if you modify one file, you then have to re-compile all files that depend on that file, and then all files which depend on that file, etc. For large codebases with many branches I could still see some benefit here.

**Comment by JoshuaKGoldberg** [comment]
Given the lack of activity on this issue and the upcoming TSLint deprecation in #4534, this seems rather out of scope now. 😢
